subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,702 @@
+/*
+ * Copyright (c) 2023 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief USB-C Power Policy Engine (PE)
+ *
+ * The information in this file was taken from the USB PD
+ * Specification Revision 3.0, Version 2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/smf.h>
+#include <zephyr/usb_c/usbc.h>
+#include <zephyr/drivers/usb_c/usbc_pd.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(usbc_stack, CONFIG_USBC_STACK_LOG_LEVEL);
+
+#include "usbc_stack.h"
+
+/**
+ * @brief Initialize the Source Policy Engine layer
+ */
+void pe_src_init(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	/* Initial role of source is DFP */
+	pe_set_data_role(dev, TC_ROLE_DFP);
+
+	/* Initialize timers */
+	usbc_timer_init(&pe->pd_t_typec_send_source_cap, PD_T_TYPEC_SEND_SOURCE_CAP_MIN_MS);
+	usbc_timer_init(&pe->pd_t_ps_hard_reset, PD_T_PS_HARD_RESET_MAX_MS);
+
+	/* Goto startup state */
+	pe_set_state(dev, PE_SRC_STARTUP);
+}
+
+/**
+ * @brief Handle source-specific DPM requests
+ */
+bool source_dpm_requests(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	if (pe->dpm_request > REQUEST_TC_END) {
+		atomic_set_bit(pe->flags, PE_FLAGS_DPM_INITIATED_AMS);
```

Comment:
It seems like this bit will be set even if we didn't change states to start a new AMS?